beacon-chain/state/types.go

```diff
@@ -110,6 +135,13 @@ func (b *BeaconState) Copy() *BeaconState {
 		}
 	}
 
+	// Finalizer runs when dst is being destroyed in garbage collection.
+	runtime.SetFinalizer(dst, func(b *BeaconState) {
+		for _, v := range b.sharedFieldReferences {
+			v.refs--
```

Comment:
could this go underflow?

Reply:
Not as it is implemented now. If it did have an underflow, it just means it would go back to the previous always copy-on-write so the risk is low